### PR TITLE
Feature reporting

### PR DESCRIPTION
**Pull request by dmolineus**

### CODE REVIEW

File: src/Version10/ToolReportInterface.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * Add an diagnostic entry.
+     *
+     * @param string      $severity The severity of the error - one of the self::SEVERITY_* constants.
+     * @param string      $message  The error message.
+     * @param string|null $file     The file of the error.
+     * @param int|null    $line     The line the error is on (within the file).
+     * @param int|null    $column   The column the error is on (within the line).
+     * @param string|null $source   The optional source of the error.
+     */
```
Comment (dmolineus) [inline src/Version10/ToolReportInterface.php:26]: We should document the exception which has to be thrown here:

```php
/** @throws ReportClosedException When report is closed. */
```

File: src/Version10/ToolReportInterface.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * Add an attachment to the tool report.
+     *
+     * @param string      $filePath The absolute path to the attachment file.
+     * @param string|null $name     The internal name of the attachment. Defaults to the basename of the passed file.
+     *
+     * @return void
+     */
+    public function addAttachment(string $filePath, ?string $name = null): void;
```
Comment (dmolineus) [inline src/Version10/ToolReportInterface.php:44]: We should document the exception which has to be thrown here:

```php
/** @throws ReportClosedException When report is closed. */
```

File: src/Version10/ToolReportInterface.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * Add an attachment to the tool report.
+     *
+     * @param string $buffer The file buffer.
+     * @param string $name   The internal name of the attachment.
+     *
+     * @return void
+     */
+    public function addBufferAsAttachment(string $buffer, string $name): void;
```
Comment (dmolineus) [inline src/Version10/ToolReportInterface.php:54]: We should document the exception which has to be thrown here:

```php
/** @throws ReportClosedException When report is closed. */
```

File: src/Version10/ToolReportInterface.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * Closes the report with the passed status.
+     *
+     * @param string $status The status of the report (either self::STATUS_PASSED or self::STATUS_FAILED).
+     *
+     * @throws ReportClosedException When the report has been closed previously.
+     */
+    public function finish(string $status): void;
```
Comment (dmolineus) [inline src/Version10/ToolReportInterface.php:63]: We have a `ReportClosedException`, but the method is called `finish`. Maybe be should use the same verb here.

File: src/Version10/Util/CheckstyleReportAppender.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                    self::getXmlAttribute($errorNode, 'message', ''),
+                    $fileName,
+                    self::getIntXmlAttribute($errorNode, 'line'),
+                    self::getIntXmlAttribute($errorNode, 'column'),
+                    self::getXmlAttribute($errorNode, 'source')
+                );
+            }
+        }
+    }
+}
```
Comment (dmolineus) [inline src/Version10/Util/CheckstyleReportAppender.php:62]: We should unify the file endings. I prefer adding a new line.

File: src/Version10/Util/BufferedLineReader.php [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        if (false === ($eol = strpos($this->data, "\n"))) {
+            return null;
+        }
+
+        $line = substr($this->data, 0, $eol);
+        $this->data = substr($this->data, $eol + 1);
+
+        return $trim ? trim($line) : $line;
+    }
+}
```
Comment (dmolineus) [inline src/Version10/Util/BufferedLineReader.php:36]: We should unify the file endings. I prefer adding a new line.

File: src/Version10/Util/XmlReportAppenderTrait.php [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    private function getIntXmlAttribute(DOMElement $element, string $attribute): ?int
+    {
+        $value = $this->getXmlAttribute($element, $attribute);
+        if ($value === null) {
+            return null;
+        }
+
+        return (int) $value;
+    }
+}
```
Comment (dmolineus) [inline src/Version10/Util/XmlReportAppenderTrait.php:120]: We should unify the file endings. I prefer adding a new line.

File: src/Version10/Util/XmlReportAppenderTrait.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    private function getXmlAttribute(DOMElement $element, string $attribute, ?string $defaultValue = null): ?string
+    {
+        if ($element->hasAttribute($attribute)) {
+            return $element->getAttribute($attribute);
+        }
+
+        return $defaultValue;
+    }
+
+    private function getIntXmlAttribute(DOMElement $element, string $attribute): ?int
```
Comment (dmolineus) [inline src/Version10/Util/XmlReportAppenderTrait.php:111]: Supporting a default value might make sense here as well. The trait is designed to be reused.

File: src/Version10/Util/JUnitReportAppender.php [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+                $this->getXmlAttribute($childNode, 'type')
+            );
+        }
+    }
+
+    private function stripRootDir(string $content): string
+    {
+        return str_replace($this->rootDir, '', $content);
+    }
+}
```
Comment (dmolineus) [inline src/Version10/Util/JUnitReportAppender.php:128]: We should unify the file endings. I prefer adding a new line.

File: src/Version10/Util/JUnitReportAppender.php [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            /*
+             * <failure> has the following attributes:
+             * - type: name of exception
+             *
+             * Text content holds error message.
+             */
+            $report->addDiagnostic(
+                $severity,
+                $this->stripRootDir($childNode->nodeValue),
+                $this->stripRootDir($testCase->getAttribute('file')),
```
Comment (dmolineus) [inline src/Version10/Util/JUnitReportAppender.php:113]: Phpspec omits the `file` attribute but provides the `classname` attribute. I'd like to resolve the file name but I'm not sure how as the autoload file is not loaded in the context, so the reflection API does not work. @discordier Do you have any idea?

Should we add a classname attribute to our `diagnostice` node as well?

File: src/Version10/Util/JUnitReportAppender.php [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+            switch ($childNode->nodeName) {
+                case 'error':
+                case 'failure':
+                    $severity = ToolReportInterface::SEVERITY_ERROR;
+                    break;
+                case 'warning':
+                    $severity = ToolReportInterface::SEVERITY_WARNING;
+                    break;
+                default:
```
[the remark this answers is not in the document]
Comment (discordier) [reply src/Version10/Util/JUnitReportAppender.php:100]: As the FIXME denotes, this is to be removed but added as a marker to detect which "common nodes" are presently used.

File: src/Version10/Util/JUnitReportAppender.php (hunk 10 above)
Comment (discordier) [reply src/Version10/Util/JUnitReportAppender.php]: Having a class in the diagnostics might come in handy.
However, I think this should rather be an array of class names as there might be multiple classes affected (as we have with files and ranges).

**Pull request merged**